### PR TITLE
document the support for constraints

### LINKED ISSUE #290: Constraints stack needs documentation
Need clear docs for the constraints stack support, including update the comparison table with SPEC commands.

### LINKED ISSUE #301: BUG: restore(configuration) should push constraints to stack
The `DiffractometerConfiguration.restore()` method should *push* (`apply_constraints()`) to the existing stack instead of replacing the stack entirely.  This preserves `undo_last_constraints()` and `reset_constraints()`.

### PR DESCRIPTION
**Pull request by prjemian**
- close #290
- close #301

### CODE REVIEW

**Comment by prjemian** [comment]
During testing (and confirming user reports), the `value` and `fit` parameters have no effect on the forward solutions.  This needs further testing.

Marking this PR as draft for now.

**Comment by prjemian** [comment]
The diffractometer (engine) `mode` is used to hold an axis at a constant value for a `forward(hkl)` calculation.  The Constraints `value` attribute:

> Constant value to be used if held constant by forward mode.

**Comment by prjemian** [comment]
This [comment](https://repo.or.cz/hkl.git/blob/HEAD:/Documentation/hkl.org.in#l22) describes the meaning of `fit`, but not in the context of forward() computations.

```
- Crystal lattice refinement
  - with more than 2 reflections you can select which parameter must
    be fitted.
```

Here are the places in *libhkl* source where `fit_get` appears and this parameter might be used:

```bash
(base) prjemian@arf:~/.../picca/hkl$ git grep fit_get
gui/hkl-gui.c:          fit = hkl_parameter_fit_get(p);                         \
gui/hkl-gui.c:          gboolean fit = hkl_parameter_fit_get(p);                \
hkl.h:HKLAPI int hkl_parameter_fit_get(const HklParameter *self) HKL_ARG_NONNULL(1);
hkl/hkl-parameter.c: * hkl_parameter_fit_get:
hkl/hkl-parameter.c:int hkl_parameter_fit_get(const HklParameter *self)
tests/hkl-axis-t.c:     ok(TRUE == hkl_parameter_fit_get(axis), __func__);
tests/hkl-axis-t.c:     ok(TRUE == hkl_parameter_fit_get(axis), __func__);
tests/hkl-axis-t.c:     ok(TRUE == hkl_parameter_fit_get(copy), __func__);
tests/hkl-axis-t.c:     ok(TRUE == hkl_parameter_fit_get(copy), __func__);
```

Not apparent it is used anywhere in the computation of `forward(hkl)` solutions.

**Comment by prjemian** [comment]
To avoid making a breaking change, by dropping this parameter altogether, consider these actions now:

- deprecate its use
- provide warning messages

**Comment by prjemian** [comment]
Confirmed with Fred Picca that the `fit` parameter is not used when calculating angles from h,k,l:

> This interface was developped for the sample parameters a/b/c/alpha/beta/gamma/ux/uy/uz in order to fit the UB martix.
> The parameter interface define a fit_get/set and since the axes derived from this interface, the  functions are available, but of no use.

**Comment by prjemian** [comment]
@mrakitin @rodolakis @strempfer @padraic-shafer -- This is ready for review now.

File: hkl/tests/test_diffract.py [hunk 1]
```diff
@@ -165,10 +165,16 @@ def test_forward_solutions_table(fourc):
...
     sol = fourc.forward(1, 0, 0)
-    assert pytest.approx(sol.omega, 1e-5) == -30
-    assert pytest.approx(sol.chi, 1e-5) == 0
-    assert pytest.approx(sol.phi, 1e-5) == -90
-    assert pytest.approx(sol.tth, 1e-5) == -60
+
+    def check_limits_and_value(axis, value, tol=1e-5):
+        assert min(fourc.calc[axis].limits) <= value
+        assert max(fourc.calc[axis].limits) >= value
+        assert pytest.approx(getattr(sol, axis), 1e-5) == value
```
Comment (padraic-shafer) [inline hkl/tests/test_diffract.py:172]:
```suggestion
        assert pytest.approx(getattr(sol, axis), rel=tol) == value
```

I believe this was intended to use the parameter, `tol` of the enclosing function.

I think it is worth explicitly [naming the tolerance parameter](https://docs.pytest.org/en/7.1.x/reference/reference.html#pytest-approx) to `pytest.approx()` which accepts both `rel`ative and `abs`olute tolerances.

**Review by padraic-shafer — APPROVED** [review]
The updated documentation is helpful and clearer. Thanks for that!

(Btw,  I skipped the SPEC docs in this review)

File: hkl/tests/test_diffract.py (hunk 1 above)
Comment (prjemian) [reply hkl/tests/test_diffract.py:172]: Good catch.  In this context, `abs` is the preference.  5 digits, 10 microdegrees.  `rel` would evaluate to more digits for angles close to zero.

Comment (prjemian) [reply hkl/tests/test_diffract.py]: Also applied to four other instances in this file.

**Review by mrakitin — APPROVED** [review]
Looks good!
I have a suggestion about the .ipynb files rendering, which can be addressed in a separate PR.

**Pull request merged**